File: frontend/src/api/insights.ts
import { api } from './client'

export interface Insight {
  id: string
  kind: string
  title: string
  body: string
  metric_a: string
  metric_b: string
  r: number
  n: number
  lag: number
  score: number
  status: 'new' | 'kept' | 'dismissed'
  feedback: number | null
  created_at: string
}

export interface BriefingToday {
  status: 'ready' | 'not_generated'
  briefing?: {
    id: string
    date: string
    content_md: string
    facts: Record<string, unknown>
    created_at: string
  }
}

export interface BriefingPreferences {
  enabled: boolean
  deliver_at: string
  channels: { push?: boolean; email?: boolean }
  tz: string
}

export interface HeatmapData {
  days: Record<string, number>
  streak: number
}

export const insightsApi = {
  discoveries: () => api.get<Insight[]>('/insights/discoveries').then(r => r.data),
  feedback: (id: string, feedback: 1 | -1) =>
    api.post<Insight>(`/insights/discoveries/${id}`, { feedback }).then(r => r.data),
  briefingToday: () => api.get<BriefingToday>('/insights/briefing/today').then(r => r.data),
  briefingPreferences: () => api.get<BriefingPreferences>('/insights/briefing/preferences').then(r => r.data),
  updateBriefingPreferences: (prefs: BriefingPreferences) =>
    api.post<BriefingPreferences>('/insights/briefing/preferences', prefs).then(r => r.data),
  heatmap: (days = 180) => api.get<HeatmapData>('/insights/heatmap', { params: { days } }).then(r => r.data),
}

File: backend/app/models/insights.py
import uuid
import datetime as dt
from typing import Optional, Dict, Any

from sqlmodel import SQLModel, Field, Column
from sqlalchemy.dialects.postgresql import JSONB

def _utcnow() -> dt.datetime:
    return dt.datetime.utcnow()

class BriefingPreference(SQLModel, table=True):
    __tablename__ = "briefing_preferences"

    user_id: uuid.UUID = Field(primary_key=True, foreign_key="users.id")
    enabled: bool = Field(default=True, nullable=False)
    deliver_at: dt.time = Field(nullable=False)
    channels: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSONB))
    tz: str = Field(default="UTC", nullable=False)


class Briefing(SQLModel, table=True):
    __tablename__ = "briefings"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    user_id: uuid.UUID = Field(foreign_key="users.id", index=True, nullable=False)
    date: dt.date = Field(nullable=False, index=True)
    content_md: str = Field(nullable=False)
    facts: Dict[str, Any] = Field(default_factory=dict, sa_column=Column(JSONB))
    created_at: dt.datetime = Field(default_factory=_utcnow, nullable=False)


class Insight(SQLModel, table=True):
    __tablename__ = "insights"

    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    user_id: uuid.UUID = Field(foreign_key="users.id", index=True, nullable=False)
    kind: str = Field(default="correlation", nullable=False)
    title: str = Field(nullable=False)
    body: str = Field(nullable=False)
    metric_a: str = Field(nullable=False)
    metric_b: str = Field(nullable=False)
    r: float = Field(nullable=False)
    n: int = Field(nullable=False)
    lag: int = Field(default=0, nullable=False)
    score: float = Field(nullable=False)
    status: str = Field(default="new", nullable=False)  # 'new', 'kept', 'dismissed'
    feedback: Optional[int] = Field(default=None)
    created_at: dt.datetime = Field(default_factory=_utcnow, nullable=False)
